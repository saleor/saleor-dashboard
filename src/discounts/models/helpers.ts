import {
  DecimalFilterInput,
  PromotionRuleDetailsFragment,
  PromotionRuleInput,
  RewardValueTypeEnum,
} from "@dashboard/graphql";

import { CataloguePredicateAPI, OrderPredicateAPI } from "../types";
import { Condition, ConditionType, isTuple } from "./Condition";
import { Rule } from "./Rule";

export const createBaseAPIInput = (data: Rule): PromotionRuleInput => {
  return {
    name: data.name,
    description: data.description ? JSON.parse(data.description) : null,
    channels: data?.channel ? [data.channel.value] : [],
    rewardType: data.rewardType,
    rewardValue: data.rewardValue,
    rewardValueType: data.rewardValueType,
  };
};

export const createBaseRuleInputFromAPI = (
  data: PromotionRuleDetailsFragment,
  giftsLabels: Record<string, string>,
): Omit<Rule, "conditions"> => {
  return {
    id: data.id,
    name: data.name ?? "",
    description: data.description ? JSON.stringify(data.description) : "",
    // For now Dashboard supports only one channel per rule
    // due to API product variant filtering limitations
    // TODO: Add support for multiple channels
    channel: data?.channels?.length
      ? { label: data?.channels[0].name, value: data?.channels[0].id }
      : null,
    rewardType: data?.rewardType ?? null,
    rewardValue: data.rewardValue ?? null,
    rewardGifts:
      data.giftIds?.map(id => ({
        value: id,
        label: giftsLabels[id],
      })) ?? [],
    rewardValueType: data.rewardValueType ?? RewardValueTypeEnum.FIXED,
  };
};

export const createAPIWhereInput = (condition: Condition) => {
  const label = condition.type;
  const value = condition.value;

  if (label === "lower") {
    return { range: { lte: value } };
  }
  if (label === "greater") {
    return { range: { gte: value } };
  }

  if (isTuple(value) && label === "between") {
    const [gte, lte] = value;
    return { range: { lte, gte } };
  }

  return { eq: value };
};

export function getConditionType(
  conditionValue: DecimalFilterInput,
): ConditionType {
  if (conditionValue.eq) {
    return "is";
  }

  if (conditionValue.range) {
    if (conditionValue.range.lte && conditionValue.range.gte) {
      return "between";
    }

    if (conditionValue.range.lte) {
      return "lower";
    }

    if (conditionValue.range.gte) {
      return "greater";
    }
  }

  return "is";
}

export function getConditionValue(conditionValue: DecimalFilterInput) {
  if (conditionValue.eq) {
    return conditionValue.eq;
  }

  if (conditionValue.oneOf) {
    return conditionValue.oneOf;
  }

  if (conditionValue.range) {
    if (conditionValue.range.lte && conditionValue.range.gte) {
      return [conditionValue.range.gte, conditionValue.range.lte];
    }

    if (conditionValue.range.lte) {
      return conditionValue.range.lte;
    }

    if (conditionValue.range.gte) {
      return conditionValue.range.gte;
    }
  }

  return conditionValue.eq;
}

const ALLOW_KEYS = ["ids", "eq", "oneOf", "range"];

export function hasPredicateNestedConditions(
  predicate: OrderPredicateAPI | CataloguePredicateAPI,
): boolean {
  const keys = Object.keys(predicate);
  if (keys.includes("AND") || keys.length > 2) {
    return true;
  }

  if (
    keys.length === 1 &&
    keys[0] !== "OR" &&
    keys[0] !== "discountedObjectPredicate"
  ) {
    const innerKeys = Object.keys(
      predicate[keys[0] as keyof typeof predicate] ?? {},
    );
    return innerKeys.every(key => !ALLOW_KEYS.includes(key));
  }

  if (predicate.OR && predicate.OR?.some(checkDeeplyNestedPredicate)) {
    return true;
  }

  if ("discountedObjectPredicate" in predicate) {
    return hasPredicateNestedConditions(predicate.discountedObjectPredicate);
  }

  return false;
}

function checkDeeplyNestedPredicate(
  nestedPredicate: OrderPredicateAPI | CataloguePredicateAPI,
): boolean {
  const keys = Object.keys(nestedPredicate);
  if (keys.includes("AND") || keys.includes("OR")) {
    return true;
  }

  type keyType = keyof typeof nestedPredicate;

  for (const key in nestedPredicate) {
    if (ALLOW_KEYS.includes(key)) {
      continue;
    }

    const innerKeys = Object.keys(nestedPredicate[key as keyType] ?? {});
    const hasNotAllowedKeys = innerKeys.every(key => !ALLOW_KEYS.includes(key));

    if (hasNotAllowedKeys) {
      return true;
    }

    if (typeof nestedPredicate[key as keyType] === "object") {
      return checkDeeplyNestedPredicate(
        nestedPredicate[key as keyType] as
          | OrderPredicateAPI
          | CataloguePredicateAPI,
      );
    }
  }
  return false;
}
