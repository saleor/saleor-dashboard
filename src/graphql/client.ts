// DON'T TOUCH THIS
// These are separate clients and do not share configs between themselves
import { ApolloClient, ApolloLink, InMemoryCache } from "@apollo/client";
import { createFetch, createSaleorClient } from "@saleor/sdk";
import { createUploadLink } from "apollo-upload-client";

import { getApiUrl } from "../config";
import introspectionQueryResultData from "./fragmentTypes.generated";
import { TypedTypePolicies } from "./typePolicies.generated";

const attachVariablesLink = new ApolloLink((operation, forward) =>
  forward(operation).map(data => ({
    ...data,
    extensions: {
      ...data.extensions,
      variables: operation.variables,
    },
  })),
);

export const link = attachVariablesLink.concat(
  createUploadLink({
    credentials: "include",
    uri: getApiUrl(),
    fetch: createFetch(),
  }) as unknown as ApolloLink, // type mismatch between apollo-upload-client and @apollo/cient
);

export const apolloClient = new ApolloClient({
  cache: new InMemoryCache({
    possibleTypes: introspectionQueryResultData.possibleTypes,
    typePolicies: {
      CountryDisplay: {
        keyFields: ["code"],
      },
      Money: {
        merge: false,
      },
      TaxedMoney: {
        merge: false,
      },
      Weight: {
        merge: false,
      },
      Shop: {
        keyFields: [],
      },
      AttributeValue: {
        fields: {
          /**
           * Since, API sometimes creates an empty slug,
           * We need to handle that case also on front-end,
           * so after fix that problem in the API, the UI will ablle
           * to handle it.
           *
           * If the slug is empty, use the name
           */
          slug: (givenSlug, { readField }) => {
            if (!givenSlug) {
              return readField("name");
            }

            return givenSlug;
          },
        },
      },
    } as TypedTypePolicies,
  }),
  link,
});

export const saleorClient = createSaleorClient({
  apiUrl: getApiUrl(),
  channel: "",
});
