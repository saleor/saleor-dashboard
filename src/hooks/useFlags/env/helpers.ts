import camelCase from "lodash/camelCase";
import snakeCase from "lodash/snakeCase";

import { ENV_FLAG_PREFIX } from "./const";

export const envNameToFlagName = (envName: string) => {
  const name = envName.split(ENV_FLAG_PREFIX)[1];
  return camelCase(name);
};

export const flagNameToEnvName = (flagName: string) =>
  `${ENV_FLAG_PREFIX}${snakeCase(flagName).toUpperCase()}`;

/**
 Referencing an virtual constant FLAGS, prepared by Vite. It populates env-based feature flags into client-side, under the virtual property FLAGS,
 Please do not use FLAGS constant directly anywhere.
*/
export const readFlagFromEnv = (flagName: string): string | undefined => {
  if (FLAGS) {
    return FLAGS[flagName];
  }

  return undefined;
};

export const readAllFlagsFromEnv = (): Record<string, string> => {
  if (FLAGS) {
    return FLAGS;
  }

  return {};
};

export const isFlagEnabled = (flag: string) => flag !== "" && flag !== "false";
