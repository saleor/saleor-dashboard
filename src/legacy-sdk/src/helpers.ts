export const isInternalToken = (owner: string): boolean => owner === "saleor";
