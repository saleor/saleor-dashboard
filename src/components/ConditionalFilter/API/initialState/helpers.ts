import { ApolloQueryResult } from "@apollo/client";
import {
  _GetChannelOperandsQuery,
  _SearchAttributeOperandsQuery,
  _SearchCategoriesOperandsQuery,
  _SearchCollectionsOperandsQuery,
  _SearchProductTypesOperandsQuery,
} from "@dashboard/graphql";

import { createBooleanOptions } from "../../constants";
import { createOptionsFromAPI } from "../Handler";
import { InitialState } from "../InitialStateResponse";
import { InitialAPIResponse } from "./types";

const isChannelQuery = (
  query: InitialAPIResponse,
): query is ApolloQueryResult<_GetChannelOperandsQuery> =>
  "channels" in query.data;

const isCollectionQuery = (
  query: InitialAPIResponse,
): query is ApolloQueryResult<_SearchCollectionsOperandsQuery> =>
  "collections" in query.data;

const isCategoryQuery = (
  query: InitialAPIResponse,
): query is ApolloQueryResult<_SearchCategoriesOperandsQuery> =>
  "categories" in query.data;

const isProductTypeQuery = (
  query: InitialAPIResponse,
): query is ApolloQueryResult<_SearchProductTypesOperandsQuery> =>
  "productTypes" in query.data;

const isAttributeQuery = (
  query: InitialAPIResponse,
): query is ApolloQueryResult<_SearchAttributeOperandsQuery> =>
  "attributes" in query.data;

export const createInitialStateFromData = (
  data: InitialAPIResponse[],
  channel: string[],
) =>
  data.reduce<InitialState>(
    (acc, query) => {
      if (isChannelQuery(query)) {
        return {
          ...acc,
          channel: (query.data?.channels ?? [])
            .filter(({ slug }) => channel.includes(slug))
            .map(({ id, name, slug }) => ({ label: name, value: id, slug })),
        };
      }

      if (isCollectionQuery(query)) {
        return {
          ...acc,
          collection: createOptionsFromAPI(
            query.data?.collections?.edges ?? [],
          ),
        };
      }

      if (isCategoryQuery(query)) {
        return {
          ...acc,
          category: createOptionsFromAPI(query.data?.categories?.edges ?? []),
        };
      }

      if (isProductTypeQuery(query)) {
        return {
          ...acc,
          productType: createOptionsFromAPI(
            query.data?.productTypes?.edges ?? [],
          ),
        };
      }

      if (isAttributeQuery(query)) {
        return {
          ...acc,
          attribute:
            query.data?.attributes?.edges.reduce(
              (acc, { node }) => ({
                ...acc,
                [node.slug ?? ""]: {
                  choices:
                    node.inputType === "BOOLEAN"
                    ? createBooleanOptions()
                    : createOptionsFromAPI(node.choices?.edges ?? []),
                  slug: node?.slug,
                  value: node?.id,
                  label: node?.name,
                  inputType: node?.inputType,
                },
              }),
              {},
            ) ?? {},
        };
      }

      return acc;
    },
    {
      channel: [],
      collection: [],
      category: [],
      productType: [],
      isAvailable: createBooleanOptions(),
      isPublished: createBooleanOptions(),
      isVisibleInListing: createBooleanOptions(),
      hasCategory: createBooleanOptions(),
      giftCard: createBooleanOptions(),
      attribute: {},
    },
  );
