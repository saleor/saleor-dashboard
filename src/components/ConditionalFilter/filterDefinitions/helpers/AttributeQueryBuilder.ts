import { AttributeInput, AttributeInputTypeEnum } from "@dashboard/graphql";

import { FilterElement } from "../../FilterElement";
import { isItemOption, isItemOptionArray, isTuple } from "../../FilterElement/ConditionValue";

export class AttributeQueryBuilder {
  static build(element: FilterElement): AttributeInput {
    const attributeSlug = element.selectedAttribute?.value;

    if (!attributeSlug) {
      return { slug: "" };
    }

    const selected = element.condition.selected;
    const inputType = element.selectedAttribute?.type as AttributeInputTypeEnum;

    if (!selected.conditionValue) {
      return { slug: attributeSlug };
    }

    const { label, type } = selected.conditionValue;
    const { value } = selected;

    if (inputType === AttributeInputTypeEnum.REFERENCE) {
      if (isItemOption(value)) {
        return { slug: attributeSlug, valueNames: [value.label] };
      }

      if (isItemOptionArray(value)) {
        if (value.length === 0) {
          return { slug: attributeSlug };
        }

        return {
          slug: attributeSlug,
          valueNames: value.map(item => item.label),
        };
      }

      return { slug: attributeSlug };
    }

    if (inputType === AttributeInputTypeEnum.BOOLEAN) {
      return {
        slug: attributeSlug,
        boolean: isItemOption(value) ? value.value === "true" : value === "true",
      };
    }

    if (label === "lower" && typeof value === "string") {
      return { slug: attributeSlug, ...this.getQueryPartByType(value, type, "lte") };
    }

    if (label === "greater" && typeof value === "string") {
      return { slug: attributeSlug, ...this.getQueryPartByType(value, type, "gte") };
    }

    if (isTuple(value) && label === "between") {
      return { slug: attributeSlug, ...this.getRangeQueryPartByType(value, type) };
    }

    if (isItemOption(value)) {
      return { slug: attributeSlug, values: [value.originalSlug || value.value] };
    }

    if (isItemOptionArray(value)) {
      return { slug: attributeSlug, values: value.map(x => x.originalSlug || x.value) };
    }

    if (typeof value === "string") {
      return { slug: attributeSlug, values: [value] };
    }

    if (Array.isArray(value) && typeof value[0] === "string") {
      return { slug: attributeSlug, values: value as string[] };
    }

    return { slug: attributeSlug };
  }

  private static getQueryPartByType(value: string, type: string, what: "lte" | "gte") {
    switch (type) {
      case "datetime":
        return { dateTime: { [what]: value } };
      case "date":
        return { date: { [what]: value } };
      default:
        return { valuesRange: { [what]: parseFloat(value) } };
    }
  }

  private static getRangeQueryPartByType(value: [string, string], type: string) {
    const [gte, lte] = value;

    switch (type) {
      case "datetime.range":
        return { dateTime: { lte, gte } };
      case "date.range":
        return { date: { lte, gte } };
      default:
        return { valuesRange: { lte: parseFloat(lte), gte: parseFloat(gte) } };
    }
  }
}
