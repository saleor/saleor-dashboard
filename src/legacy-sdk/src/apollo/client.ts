import {
  ApolloClient,
  createHttpLink,
  FetchResult,
  InMemoryCache,
  NormalizedCacheObject,
  Reference,
} from "@apollo/client";
import fetch from "cross-fetch";
import jwtDecode from "jwt-decode";

import { JWTToken } from "../core";
import { auth, AuthSDK } from "../core/auth";
import { storage } from "../core/storage";
import { isInternalToken } from "../helpers";
import { TypedTypePolicies } from "./apollo-helpers";
import { ExternalRefreshMutation, RefreshTokenMutation } from "./types";

let client: ApolloClient<NormalizedCacheObject>;
let authClient: AuthSDK;
let refreshPromise:
  | ReturnType<AuthSDK["refreshToken"]>
  | ReturnType<AuthSDK["refreshExternalToken"]>
  | null = null;
const isTokenRefreshExternal = (
  result: RefreshTokenMutation | ExternalRefreshMutation,
): result is ExternalRefreshMutation => "externalRefresh" in result;

export type FetchConfig = Partial<{
  /**
   * Enable auto token refreshing. Default to `true`.
   */
  autoTokenRefresh: boolean;
  /**
   * Set a value for skew between local time and token expiration date in
   * seconds (only together with `autoTokenRefresh`). Defaults to `120`.
   */
  tokenRefreshTimeSkew: number;
  /**
   * Refresh token and retry the request when Saleor responds with `Unauthorized` error.
   * Defaults to `true`.
   */
  refreshOnUnauthorized: boolean;
}>;

export const createFetch =
  ({
    autoTokenRefresh = true,
    tokenRefreshTimeSkew = 120,
    refreshOnUnauthorized = true,
  }: FetchConfig = {}) =>
  async (input: RequestInfo, init: RequestInit = {}): Promise<Response> => {
    if (!client) {
      throw new Error(
        "Could not find Saleor's client instance. Did you forget to call createSaleorClient()?",
      );
    }

    let token = storage.getAccessToken();

    try {
      if (
        ["refreshToken", "externalRefresh"].includes(
          // INFO: Non-null assertion is enabled because the block is wrapped inside try/catch
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          JSON.parse(init.body!.toString()).operationName,
        )
      ) {
        return fetch(input, init);
      }
    } catch (e) {
      // no-op
    }

    if (autoTokenRefresh && token) {
      // auto refresh token before provided time skew (in seconds) until it expires
      const decodedToken = jwtDecode<JWTToken>(token);
      const expirationTime = (decodedToken.exp - tokenRefreshTimeSkew) * 1000;
      const owner = decodedToken.owner;

      try {
        if (refreshPromise) {
          await refreshPromise;
        } else if (Date.now() >= expirationTime) {
          if (isInternalToken(owner)) {
            await authClient.refreshToken();
          } else {
            await authClient.refreshExternalToken();
          }
        }
      } catch (e) {
        // no-op
      } finally {
        refreshPromise = null;
      }

      token = storage.getAccessToken();
    }

    if (token) {
      init.headers = {
        ...init.headers,
        "authorization-bearer": token,
      };
    }

    if (refreshOnUnauthorized && token) {
      const response = await fetch(input, init);
      const data: FetchResult = await response.clone().json();
      const isUnauthenticated = data?.errors?.some(
        error => error.extensions?.exception.code === "ExpiredSignatureError",
      );
      let refreshTokenResponse: FetchResult<
        RefreshTokenMutation | ExternalRefreshMutation,
        Record<string, unknown>,
        Record<string, unknown>
      > | null = null;
      const owner = jwtDecode<JWTToken>(token).owner;

      if (isUnauthenticated) {
        try {
          if (refreshPromise) {
            refreshTokenResponse = await refreshPromise;
          } else {
            refreshPromise = isInternalToken(owner)
              ? authClient.refreshToken()
              : authClient.refreshExternalToken();
            refreshTokenResponse = await refreshPromise;
          }

          if (
            refreshTokenResponse.data && isTokenRefreshExternal(refreshTokenResponse.data)
              ? refreshTokenResponse.data.externalRefresh?.token
              : refreshTokenResponse.data?.tokenRefresh?.token
          ) {
            // check if mutation returns a valid token after refresh and retry the request
            return createFetch({
              autoTokenRefresh: false,
              refreshOnUnauthorized: false,
            })(input, init);
          } else {
            // after Saleor returns ExpiredSignatureError status and token refresh fails
            // we log out the user and return the failed response
            authClient.logout();
          }
        } catch (e) {
          // no-op
        } finally {
          refreshPromise = null;
        }
      }

      return response;
    }

    return fetch(input, init);
  };

const getTypePolicies = (autologin: boolean): TypedTypePolicies => ({
  Query: {
    fields: {
      authenticated: {
        read(_, { readField, toReference }): boolean {
          return !!readField(
            "id",
            toReference({
              __typename: "User",
            }),
          );
        },
      },
      me: {
        read(_, { toReference, canRead }): Reference | undefined | null {
          const ref = toReference({
            __typename: "User",
          });

          return canRead(ref) ? ref : null;
        },
      },
      authenticating: {
        read(read = autologin && !!storage.getRefreshToken(), { readField }): boolean {
          if (readField("authenticated")) {
            return false;
          }

          return read;
        },
      },
    },
  },
  User: {
    /**
     * IMPORTANT
     * This works as long as we have 1 User cache object which is the current logged in User.
     * If the client should ever fetch additional Users, this should be removed
     * and the login methods (token create or verify) should be responsible for writing USER query cache manually.
     */
    keyFields: [],
    fields: {
      addresses: {
        merge: false,
      },
    },
  },
});

export const createApolloClient = (
  apiUrl: string,
  autologin: boolean,
  fetchOptions?: FetchConfig,
): ApolloClient<NormalizedCacheObject> => {
  const httpLink = createHttpLink({
    fetch: createFetch(fetchOptions),
    uri: apiUrl,
    credentials: "include",
  });

  const cache = new InMemoryCache({
    typePolicies: getTypePolicies(autologin),
  });

  client = new ApolloClient({
    cache,
    link: httpLink,
  });

  /**
   * Refreshing token code should stay under core/auth.ts To get this method available,
   * we need to call "auth()" here. refreshToken mutation doesn't require channel, so it
   * doesn't have to be populated with value.
   */
  authClient = auth({ apolloClient: client, channel: "" });

  return client;
};
