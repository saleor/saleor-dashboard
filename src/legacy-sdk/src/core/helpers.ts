import { Maybe, UserPermission } from "../apollo/types";

export const hasNonEmptyPermissions = (
  permissions: Maybe<Array<Maybe<Pick<UserPermission, "code" | "name">>>> | undefined,
): boolean => (permissions ? permissions.length > 0 : false);
