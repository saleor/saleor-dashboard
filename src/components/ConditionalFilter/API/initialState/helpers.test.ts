import { ApolloQueryResult } from "@apollo/client";
import {
  _GetChannelOperandsQuery,
  _SearchAttributeOperandsQuery,
  _SearchCategoriesOperandsQuery,
  _SearchCollectionsOperandsQuery,
  _SearchProductTypesOperandsQuery,
} from "@dashboard/graphql";

import { createInitialStateFromData } from "./helpers";

describe("ConditionalFilter / API / createInitialStateFromData", () => {
  it("should create initial state from queries", () => {
    // Arrange
    const channelQuery = {
      data: {
        channels: [
          { id: "1", name: "Channel 1", slug: "channel-1" },
          { id: "2", name: "Channel 2", slug: "channel-2" },
        ],
      },
    } as ApolloQueryResult<_GetChannelOperandsQuery>;

    const collectionQuery = {
      data: {
        collections: {
          edges: [
            { node: { id: "1", name: "Collection 1", slug: "collection-1" } },
            { node: { id: "2", name: "Collection 2", slug: "collection-2" } },
          ],
        },
      },
    } as ApolloQueryResult<_SearchCollectionsOperandsQuery>;

    const categoryQuery = {
      data: {
        categories: {
          edges: [
            { node: { id: "1", name: "Category 1", slug: "category-1" } },
            { node: { id: "2", name: "Category 2", slug: "category-2" } },
          ],
        },
      },
    } as ApolloQueryResult<_SearchCategoriesOperandsQuery>;

    const productTypeQuery = {
      data: {
        productTypes: {
          edges: [
            {
              node: { id: "1", name: "Product Type 1", slug: "product-type-1" },
            },
            {
              node: { id: "2", name: "Product Type 2", slug: "product-type-2" },
            },
          ],
        },
      },
    } as ApolloQueryResult<_SearchProductTypesOperandsQuery>;

    const attributeQuery = {
      data: {
        attributes: {
          edges: [
            {
              node: {
                id: "1",
                name: "Attribute 1",
                slug: "attribute-1",
                inputType: "MULTISELECT",
                choices: {
                  edges: [
                    {
                      node: {
                        id: "1",
                        name: "Choice 1",
                        slug: "choice-1",
                      },
                    },
                    {
                      node: {
                        id: "2",
                        name: "Choice 2",
                        slug: "choice-2",
                      },
                    },
                  ],
                },
              },
            },
            {
              node: {
                id: "2",
                name: "Attribute 2",
                slug: "attribute-2",
                inputType: "BOOLEAN",
              },
            },
          ],
        },
      },
    } as ApolloQueryResult<_SearchAttributeOperandsQuery>;
    const data = [
      channelQuery,
      collectionQuery,
      categoryQuery,
      productTypeQuery,
      attributeQuery,
    ];
    const channel = ["channel-1"];
    // Act
    const result = createInitialStateFromData(data, channel);
    // Assert
    expect(result).toMatchSnapshot();
  });
});
