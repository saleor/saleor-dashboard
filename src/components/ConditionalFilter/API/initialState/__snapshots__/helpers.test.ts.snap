// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConditionalFilter / API / createInitialStateFromData should create initial state from queries 1`] = `
Object {
  "attribute": Object {
    "attribute-1": Object {
      "choices": Array [
        Object {
          "label": "Choice 1",
          "originalSlug": undefined,
          "slug": "choice-1",
          "value": "1",
        },
        Object {
          "label": "Choice 2",
          "originalSlug": undefined,
          "slug": "choice-2",
          "value": "2",
        },
      ],
      "inputType": "MULTISELECT",
      "label": "Attribute 1",
      "slug": "attribute-1",
      "value": "1",
    },
    "attribute-2": Object {
      "choices": Array [
        Object {
          "label": "Yes",
          "slug": "true",
          "type": undefined,
          "value": "true",
        },
        Object {
          "label": "No",
          "slug": "false",
          "type": undefined,
          "value": "false",
        },
      ],
      "inputType": "BOOLEAN",
      "label": "Attribute 2",
      "slug": "attribute-2",
      "value": "2",
    },
  },
  "category": Array [
    Object {
      "label": "Category 1",
      "originalSlug": undefined,
      "slug": "category-1",
      "value": "1",
    },
    Object {
      "label": "Category 2",
      "originalSlug": undefined,
      "slug": "category-2",
      "value": "2",
    },
  ],
  "channel": Array [
    Object {
      "label": "Channel 1",
      "slug": "channel-1",
      "value": "1",
    },
  ],
  "collection": Array [
    Object {
      "label": "Collection 1",
      "originalSlug": undefined,
      "slug": "collection-1",
      "value": "1",
    },
    Object {
      "label": "Collection 2",
      "originalSlug": undefined,
      "slug": "collection-2",
      "value": "2",
    },
  ],
  "giftCard": Array [
    Object {
      "label": "Yes",
      "slug": "true",
      "type": undefined,
      "value": "true",
    },
    Object {
      "label": "No",
      "slug": "false",
      "type": undefined,
      "value": "false",
    },
  ],
  "hasCategory": Array [
    Object {
      "label": "Yes",
      "slug": "true",
      "type": undefined,
      "value": "true",
    },
    Object {
      "label": "No",
      "slug": "false",
      "type": undefined,
      "value": "false",
    },
  ],
  "isAvailable": Array [
    Object {
      "label": "Yes",
      "slug": "true",
      "type": undefined,
      "value": "true",
    },
    Object {
      "label": "No",
      "slug": "false",
      "type": undefined,
      "value": "false",
    },
  ],
  "isPublished": Array [
    Object {
      "label": "Yes",
      "slug": "true",
      "type": undefined,
      "value": "true",
    },
    Object {
      "label": "No",
      "slug": "false",
      "type": undefined,
      "value": "false",
    },
  ],
  "isVisibleInListing": Array [
    Object {
      "label": "Yes",
      "slug": "true",
      "type": undefined,
      "value": "true",
    },
    Object {
      "label": "No",
      "slug": "false",
      "type": undefined,
      "value": "false",
    },
  ],
  "productType": Array [
    Object {
      "label": "Product Type 1",
      "originalSlug": undefined,
      "slug": "product-type-1",
      "value": "1",
    },
    Object {
      "label": "Product Type 2",
      "originalSlug": undefined,
      "slug": "product-type-2",
      "value": "2",
    },
  ],
}
`;
